etcdserver/v3_server.go

```diff
@@ -803,3 +805,101 @@ func (s *EtcdServer) AuthInfoFromCtx(ctx context.Context) (*auth.AuthInfo, error
 	return authInfo, nil
 
 }
+
+func (s *EtcdServer) Downgrade(ctx context.Context, r *pb.DowngradeRequest) (*pb.DowngradeResponse, error) {
+	switch r.Action {
+	case pb.DowngradeRequest_VALIDATE:
+		return s.downgradeValidate(ctx, r.Version)
+	case pb.DowngradeRequest_ENABLE:
+		return s.downgradeEnable(ctx, r)
+	case pb.DowngradeRequest_CANCEL:
+		return s.downgradeCancel(ctx)
+	}
+	return nil, nil
+}
+
+func (s *EtcdServer) downgradeValidate(ctx context.Context, v string) (*pb.DowngradeResponse, error) {
+	resp := &pb.DowngradeResponse{}
+	var err error
+
+	targetVersion, err := semver.NewVersion(v)
+	if err != nil {
+		return nil, fmt.Errorf("wrong version format: %v", err)
+	}
+	targetVersion = &semver.Version{Major: targetVersion.Major, Minor: targetVersion.Minor}
+
+	// do linearized read to avoid using stale downgrade information
+	err = s.linearizableReadNotify(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cv := s.ClusterVersion()
```

Comment:
could cv be nil?

Reply:
If not all members in cluster are started, server version of those 'unstarted' members are unknown. In this case we cannot determine if a downgrade target version is valid.

Reply:
Let's say we want to start a three-member cluster. If we only start one member, the cluster is not able to serve user request including downgrade request. If starting more than half members, the cluster's version is minClusterVersion which is 3.0 currently. `Downgrade.Validate` will check target version against 3.0.
But I think we can check `cv` to avoid nil pointer panic.

Reply:
Cluster version is monitored and updated every few seconds, I guess it is still possible that cv is nil when it is being used here.
If cv if nil, return code Unavailable, so that the client knows they can retry later.